taas/mapping.py

```diff
@@ -0,0 +1,99 @@
+from abc import ABCMeta, abstractmethod
+
+
+class Mapping(object):
+    """
+        This class handles mappings between our raw input, and what we generate.
+    """
+
+    # We're an abstract class
+    __metaclass__ = ABCMeta
+
+    # This was originally called 'value', but apparently that's a
+    # special method name.
+    @abstractmethod
+    def emit(self, row):
+        """
+            Given a row of data, emits the value calculated by
+            this transform.
+        """
+        raise NotImplementedError
+
+
+class Literal(Mapping):
+    """A literal string that's always the same."""
+
+    def __init__(self, config):
+        self.value = config["value"]
+
+    def emit(self, row):
+        return self.value
+
+
+class Map(Mapping):
+    """The most common mapping. Turns a field in our data into a field in our output."""
+
+    def __init__(self, config):
+        self.field = config["field"]
+        self.optional = config.get("optional", False)
+
+    def emit(self, row):
+        value = row[self.field]
+
+        if len(value) == 0:
+            if self.optional:
+                value = None
+            else:
+                raise ValueError("Required field {} missing in data".format(self.field))
+
+        return value
+
+
+class Concat(Map):
+    """A map, but concatenated with static strings."""
+
+    def __init__(self, config):
+        super(self.__class__, self).__init__(config)
+        self.pre = config.get("prefix", "")
+        self.post = config.get("postfix", "")
+
+    def emit(self, row):
+        value = super(self.__class__, self).emit(row)
+
+        # This implies our field was optional. If it wasn't,
+        # our parent class would have raised an exception.
```

Comment:
Right now as a truly awful stack backtrace.

The right place to change that is in the higher-level code that's calling this, since it's aware of the file which is being processed, the row number, and so in. We still want the exception thrown from this class; the caller can then catch and decorate it in a human friendly fashion. Eg: "Row 26 of <url> is missing required field fooble".

Opened as TAAS-33.